File: beekeeperApp/shared/src/commonMain/kotlin/com/beekeeper/app/domain/model/Task.kt
package com.beekeeper.app.domain.model

import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.Serializable

/**
 * Represents a beekeeping task or reminder
 */
@Serializable
data class BeekeepingTask(
    val id: String,
    val title: String,
    val description: String = "",
    val taskType: TaskType,

    // Scheduling
    val dueDate: LocalDateTime,
    val reminderDate: LocalDateTime? = null,
    val recurrence: TaskRecurrence? = null,

    // Association
    val hiveId: String? = null,  // null for general tasks
    val relatedInspectionId: String? = null,

    // Status
    val status: TaskStatus = TaskStatus.PENDING,
    val priority: TaskPriority = TaskPriority.MEDIUM,
    val completedDate: LocalDateTime? = null,

    // Additional info
    val estimatedDurationMinutes: Int? = null,
    val weatherDependent: Boolean = false,
    val minimumTemperature: Double? = null,  // Minimum temp for task
    val notes: String = "",
    val checklist: List<TaskChecklistItem> = emptyList(),

    val createdAt: String,
    val updatedAt: String
)

/**
 * Checklist item within a task
 */
@Serializable
data class TaskChecklistItem(
    val id: String,
    val description: String,
    val completed: Boolean = false
)

/**
 * Recurrence pattern for repeating tasks
 */
@Serializable
data class TaskRecurrence(
    val frequency: RecurrenceFrequency,
    val interval: Int = 1,  // Every X frequency units
    val endDate: LocalDateTime? = null,
    val count: Int? = null  // Number of occurrences
)

/**
 * Type of beekeeping task
 */
@Serializable
enum class TaskType {
    // Regular maintenance
    INSPECTION,
    FEEDING,
    WATER_CHECK,

    // Seasonal
    SPRING_INSPECTION,
    SUMMER_INSPECTION,
    FALL_PREPARATION,
    WINTER_CHECK,

    // Health
    PEST_TREATMENT,
    DISEASE_TREATMENT,
    MEDICATION,

    // Production
    HARVEST_HONEY,
    EXTRACT_HONEY,
    BOTTLE_HONEY,
    HARVEST_WAX,
    HARVEST_PROPOLIS,

    // Colony management
    SPLIT_HIVE,
    COMBINE_HIVES,
    REQUEEN,
    SWARM_PREVENTION,
    SWARM_COLLECTION,

    // Equipment
    ADD_BOXES,
    REMOVE_BOXES,
    CLEAN_EQUIPMENT,
    REPAIR_EQUIPMENT,
    BUILD_FRAMES,

    // Other
    EDUCATION,
    RECORD_KEEPING,
    ORDER_SUPPLIES,
    GENERAL,
    OTHER
}

/**
 * Status of a task
 */
@Serializable
enum class TaskStatus {
    PENDING,
    IN_PROGRESS,
    COMPLETED,
    CANCELLED,
    OVERDUE
}

/**
 * Priority level of a task
 */
@Serializable
enum class TaskPriority {
    LOW,
    MEDIUM,
    HIGH,
    URGENT
}

/**
 * Frequency for recurring tasks
 */
@Serializable
enum class RecurrenceFrequency {
    DAILY,
    WEEKLY,
    BIWEEKLY,
    MONTHLY,
    QUARTERLY,
    YEARLY
}

/**
 * Template for common beekeeping tasks
 */
@Serializable
data class TaskTemplate(
    val id: String,
    val name: String,
    val taskType: TaskType,
    val description: String,
    val estimatedDurationMinutes: Int,
    val defaultChecklist: List<String>,
    val seasonalRecommendation: List<Season>,
    val weatherDependent: Boolean,
    val minimumTemperature: Double?
)

/**
 * Season for seasonal recommendations
 */
@Serializable
enum class Season {
    SPRING,
    SUMMER,
    FALL,
    WINTER
}

File: api/app/models/task.py
from enum import Enum as PyEnum
from datetime import datetime
from typing import Optional

from beanie import Document

from .base import TimestampMixin


class TaskType(str, PyEnum):
    # Regular maintenance
    INSPECTION = "INSPECTION"
    FEEDING = "FEEDING"
    WATER_CHECK = "WATER_CHECK"

    # Seasonal
    SPRING_INSPECTION = "SPRING_INSPECTION"
    SUMMER_INSPECTION = "SUMMER_INSPECTION"
    FALL_PREPARATION = "FALL_PREPARATION"
    WINTER_CHECK = "WINTER_CHECK"

    # Health
    PEST_TREATMENT = "PEST_TREATMENT"
    DISEASE_TREATMENT = "DISEASE_TREATMENT"
    MEDICATION = "MEDICATION"

    # Production
    HARVEST_HONEY = "HARVEST_HONEY"
    EXTRACT_HONEY = "EXTRACT_HONEY"
    BOTTLE_HONEY = "BOTTLE_HONEY"
    HARVEST_WAX = "HARVEST_WAX"
    HARVEST_PROPOLIS = "HARVEST_PROPOLIS"

    # Colony management
    SPLIT_HIVE = "SPLIT_HIVE"
    COMBINE_HIVES = "COMBINE_HIVES"
    REQUEEN = "REQUEEN"
    SWARM_PREVENTION = "SWARM_PREVENTION"
    SWARM_COLLECTION = "SWARM_COLLECTION"

    # Equipment
    ADD_BOXES = "ADD_BOXES"
    REMOVE_BOXES = "REMOVE_BOXES"
    CLEAN_EQUIPMENT = "CLEAN_EQUIPMENT"
    REPAIR_EQUIPMENT = "REPAIR_EQUIPMENT"
    BUILD_FRAMES = "BUILD_FRAMES"

    # Other
    EDUCATION = "EDUCATION"
    RECORD_KEEPING = "RECORD_KEEPING"
    ORDER_SUPPLIES = "ORDER_SUPPLIES"
    GENERAL = "GENERAL"
    OTHER = "OTHER"


class TaskStatus(str, PyEnum):
    PENDING = "PENDING"
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"
    OVERDUE = "OVERDUE"


class TaskPriority(str, PyEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    URGENT = "URGENT"


class RecurrenceFrequency(str, PyEnum):
    DAILY = "DAILY"
    WEEKLY = "WEEKLY"
    BIWEEKLY = "BIWEEKLY"
    MONTHLY = "MONTHLY"
    QUARTERLY = "QUARTERLY"
    YEARLY = "YEARLY"


class Task(Document, TimestampMixin):
    id: str  # type: ignore[assignment]
    title: str
    description: str = ""
    task_type: TaskType = TaskType.GENERAL

    # Scheduling
    due_date: datetime
    reminder_date: Optional[datetime] = None

    # Recurrence
    recurrence_frequency: Optional[RecurrenceFrequency] = None
    recurrence_interval: Optional[int] = 1
    recurrence_end_date: Optional[datetime] = None
    recurrence_count: Optional[int] = None

    # Association
    hive_id: Optional[str] = None
    apiary_id: Optional[str] = None
    user_id: str

    # Status
    status: TaskStatus = TaskStatus.PENDING
    priority: TaskPriority = TaskPriority.MEDIUM
    completed_date: Optional[datetime] = None

    # Additional info
    estimated_duration_minutes: Optional[int] = None
    weather_dependent: bool = False
    minimum_temperature: Optional[float] = None
    notes: str = ""

    class Settings:
        name = "tasks"
        indexes = [
            "hive_id",
            "apiary_id",
            [("user_id", 1), ("due_date", 1)],
            [("user_id", 1), ("status", 1)],
        ]
